File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    stack = []
    visited = []
    for i in range(n):
        if len(visited) == n:
            break
        if i not in visited:
            stack.append(i)
            while stack:
                x = stack.pop()
                if x not in visited:
                    visited.append(x)
                for idx, computer in enumerate(computers[x]):
                    if idx not in visited and computer == 1:
                        stack.append(idx)
            answer += 1
    
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    int[] parent;
    
    int find(int x) {
        if(x == parent[x])
            return x;
        parent[x] = find(parent[x]);
        return parent[x];
    }
    
    void union(int x, int y) {
        int root1 = find(x);
        int root2 = find(y);
        parent[root1] = root2;
    }
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        parent = new int[n];
        for(int i = 0; i<n; i++) {
            parent[i] = i;
        }
        for(int i = 0; i<n; i++) {
            for(int j = 0; j<n; j++) {
                if(i==j) continue;
                if(computers[i][j] == 1 && find(i) != find(j))
                    union(i, j);
            }
        }
        HashSet<Integer> candidates = new HashSet<>();
        for(int i = 0; i<n; i++) {
            candidates.add(find(i));
        }
        return candidates.size();
    }
}